2-explicit-polymorphism/2-explicit-polymorphism.md

```diff
@@ -0,0 +1,147 @@
+# Introduction
+
+Currently in Hazel every function has to be assigned a monomorphic type.
+This makes it impossible to write code that operates on more than one type
+of data. We currently have an example that implements `map : (Num -> Num)
+-> [Num] -> [Num]`. By adopting this proposal, Hazel's map function can
+have the type `forall a. (a -> a) -> [a] -> [a]`, and thereby be useful for
+a larger set of values.
+
+Specifically, we propose extending Hazel with two type constructs
+and two expression constructs, bringing Hazel's type system up to
+a System F with holes than a Simply Typed Lambda Calculus with holes.
+
+# Type Language
+
+We'd have to extend UHTyp.t with `TyVar` and `Forall` cases:
+
+```reason
+/* UHTyp.re */
+
+type operand =
+...
+| Forall(TPat.t, t);
+| TyVar(VarErrStatus.t, Var.t)
+  ```
+
+  A type pattern is the part of the language between the `forall` and the `.`
+  in `forall a. <type>`. Type patterns will be very simple to begin with:
+
+  ```reason
+  /* TPat.t */
+  type t =
+  | Hole(MetaVar.t)
+| Var(Var.t)
+  ```
+
+  <!-- TODO: Talk about how TPat.t's might expand in the future -->
+
+# Expression Language
+
+  ```reason
+  /* UHExp.t */
+  type operand =
+  ...
+  | TyLam(TPat.t, block)
+| TyVar(TPat.t)
+  ```
+
+# Construct actions
+
+  We'll be adding the following to `Construct(shape)`:
+
+  ```
+  type shape =
+  ...
+  | SForall
+  | SType
+  ```
+
+  `Construct(SForall)` will be triggered when the user types "forall" followed by
+  a space in a type hole. This will add a Forall type with the cursor at the type
+  pattern hole and a hole as the body.
+
+  `Construct(SType)` will be triggered when the user types "type" followed by a
+  space in an expression hole. This will add a `TyVar.t` in the expression with a
+  type pattern hole.
+
+# Concrete Syntax / Representation
+
+  Forall's appear within types and look like: `forall a. ?`, where `a` is
+  a TyVar.t.
+
+  There are Type Vars types and expressions. Type Vars in types appear as a
+  string of lowercase letter without any prefix or constructor. Type Vars in
+  expressions appear as `(type a)` where `a` is also a string of lowercase
+  letters.
+
+  Type Lambdas are written like a lambda: you start with `\`. Then you write
```

Comment:
This means you have to extend the syntax of patterns too.

Reply:
What does?